lib/drivers.js

```diff
@@ -0,0 +1,157 @@
+    if (!_.isArray(platformNames)) {
+      msg += `Drivers.json entry for '${driverName}' has missing or ` +
+             `incorrect supported platformName list. `;
+    }
+
+    for (const pName of platformNames) {
+      if (!_.isString(pName)) {
+        msg += `Drivers.json entry for '${driverName}' has an incorrectly ` +
+               `formatted platformName. `;
+      }
+    }
+
+    const automationName = INSTALLED_DRIVERS[driverName].automationName;
+    if (!_.isString(automationName)) {
+      msg += `Drivers.json entry for '${driverName}' has missing ` +
+             `or incorrect automationName. `;
+    }
+    if (_.includes(automationNames, automationName)) {
+      msg += `Multiple drivers claim support for automationName ` +
+             `'${automationName}'. `;
+    }
+    automationNames.push(INSTALLED_DRIVERS[driverName].automationName);
+
+    if (!_.isString(INSTALLED_DRIVERS[driverName].version)) {
+      msg += `Drivers.json entry for '${driverName}' has missing ` +
+             `or incorrect version. `;
+    }
+
+    if (!_.isString(INSTALLED_DRIVERS[driverName].pkg)) {
+      msg += `Drivers.json entry for '${driverName}' has missing ` +
+             `or incorrect NPM package name. `;
+    }
+
+    if (!_.isString(INSTALLED_DRIVERS[driverName].mainClass)) {
+      msg += `Drivers.json entry for '${driverName}' has missing ` +
+             `or incorrect class name. `;
+    }
+  }
+
+  if (msg) {
+    throw new Error(`Appium encountered an error while trying to verify ` +
+              `the data contained in the drivers file (${DRIVERS_JSON}): ` +
+              `${msg}Please ensure you have installed drivers correctly ` +
+              `using the Appium CLI.`);
+  }
+
+  return INSTALLED_DRIVERS;
+}
+
+function printAvailableDrivers () {
+  const driverNames = Object.keys(INSTALLED_DRIVERS);
+  if (!driverNames.length) {
+    log.info('No drivers have been installed. Use the "appium driver" ' +
+             'command to install the one(s) you want to use.');
+    return;
+  }
+
+  log.info('Available drivers:');
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const ver = INSTALLED_DRIVERS[driverName].version;
+    const aName = INSTALLED_DRIVERS[driverName].automationName;
+    log.info(`  - ${driverName}@${ver} (automationName '${aName}')`);
+  }
+}
+
+function getDriverBySupport (automationName, platformName) {
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const installed = INSTALLED_DRIVERS[driverName];
+    const aNameMatches = installed.automationName.toLowerCase() ===
+                         automationName.toLowerCase();
+    const lower = (x) => x.toLowerCase();
+    const pNameMatches = _.includes(installed.platformNames.map(lower),
+                                    platformName.toLowerCase());
+
+    if (aNameMatches && pNameMatches) {
+      return {driverName, ...installed};
+    } else if (aNameMatches) {
```

Comment:
else is redundant

Reply:
👍

Reply:
actually, it's not redundant. it's handling the case where `aNameMatches && !pNameMatches`

Reply:
you can keep `if (aNameMatches)`. I was only talking about `else`

Reply:
oh i see 👍